### ISSUE TITLE
web socket causes MIN_TIME_BETWEEN_FORCED_SCANS to be enforced

### ISSUE THREAD

**Issue by keatontaylor** [keatontaylor/alexa_media_player#132]
Currently a player state change remotely will work as intended, the websocket will trigger an update/refresh methods and updates the player state on the HA GUI.

However, this is not true for users who hit the play/pause button within the HA GUI now, the state is instead left as play after hitting the play button or pause. This state will not update since if the web socket connection is established it will not poll.

Problem seems to be that the self.update() call is getting called both when the user hits play and again when the websocket message comes in, this rapid succession triggers the MIN_TIME_BETWEEN_FORCED_SCANS and prevents the player from updating accordingly. Ideally, we should only trigger refreshes that poll amazon's API from websocket events when a websocket is established for those things that can be updated from user behavior.

**Comment by keatontaylor** [comment]
So, it might make sense for us to modify the update behavior for websockets. The goal is that we want to trigger a player update when we get a player state change. Right now it seems that we're trigging based on user action, which grabs the player state which may not yet have changed and we're suppressing the second update triggered from the websocket message for player state change comes in due to the min time between forced scans being enforced. 

Did some poking around on https://alexa.amazon.com and it seems that they do no throttling when updating based on websocket messages and since player state changes are typically fairly well spread apart I think it is safe for us to remove throttling when web sockets is enabled. 

Ideally for actions that will trigger a websocket message such as play/pause/volume/etc we should avoid self.update() and just update when the websocket message is received.

**Comment by alandtse** [comment]
I agree with that approach re the suppressed second update. 

It should be fairly trivial to take the player state change to change the play/pause state even without the update. However, when I looked at the websocket messages, it didn't provide the other media player info which then required a `self.update()`. Did you see any more info we can process immediately?

So, assuming the `player_state` will always fire appropriately, I think we can only do `self.update()` only after a `player_state` change.  `last_called_change` will only update HA state with `schedule_update_ha_state` and never do a force_refresh (which implicitly calls for a `self.update()`.

On your second point about ideally not doing updates at all for websocket. I don't know if that will work unless you're only talking about `PUSH_STATE` commands. In that case I agree. `PLAYER_STATE` messages should be processed with an update as mentioned above.

**Comment by keatontaylor** [comment]
So we have the following messages from the WebSocket that are useful:

- PUSH_ACTIVITY
  - Activity from a device/account has occurred, this is currently used on https://alexa.amazon.com to trigger a rest call to update the cards displayed on the homepage.
  - This could be useful for new devices added to the Alexa account, but I have not confirmed. 
  - It is useful for updating the last_called boolean value. 

- PUSH_AUDIO_PLAYER_STATE
  - Something about the player state has changed, play/pause/seek/etc this should in all cases trigger a refresh via the get_state() AlexaAPI call.

- PUSH_VOLUME_CHANGE
  - The volume has changed or been muted. Includes the new volume state. So easy to update without a RESTful call.

- PUSH_DOPPLER_CONNECTION_CHANGE
  - State change of the devices online status, this is useful for updating the availability of the media player devices, as mentioned in the other pull request.

Currently for any player action on https://alexa.amazon.com initiated by the user, such as hitting play/pause/next/previous will not initiate a grab of the player state info from a REST request. Instead the website waits for the  PUSH_AUDIO_PLAYER_STATE websocket message and then proceeds call the playerState endpoint (in our case AlexaAPI get_state() method)

```
    def media_pause(self):
        """Send pause command."""
        if not (self.state in [STATE_PLAYING, STATE_PAUSED]
                and self.available):
            return
        self.alexa_api.pause()
        self.update()
```
In other words, don't do the self.update() above, just wait for the WebSocket PUSH_AUDIO_PLAYER_STATE message and update there.

**Comment by alandtse** [comment]
Oh you're right. Yes we probably do need the check on whether we have websockets and only issue the update there if it's not enabled. Do you want to handle?

**Comment by keatontaylor** [comment]
I'm wondering if we should even bother supporting non-websocket media player polling. I'm pretty sure that amazon has support for web sockets in every region but cannot be certain. Just don't want to cobble together a solution that tries to do both, when amazon's own website relies heavily on websockets.

**Comment by alandtse** [comment]
Let's plan to deprecate the polling in 1 or 2 releases then. I just want to avoid the complaints if Amazon opens in a new region without websockets. With the Doppler changes you added I think you're right that polling is unnecessary at this point.

**Comment by keatontaylor** [comment]
I'll check and see what the WebSocket messages if any are for adding/removing devices to an account.

**Comment by alandtse** [comment]
I have a fix in my dev branch that disables the manual update. I'm testing now.

**Comment by keatontaylor** [comment]
I did some digging into how what Websocket messages are emitted when a new device is added to an account or when a device is removed. 

Looks like PUSH_DOPPLER_CONNECTION_CHANGE is pushed when a device is added and removed. Removed devices go to offline, from what it looks like. 

So we have two options for removing polling for new devices:

1. We just trigger a update_devices() call on any PUSH_DOPPLER_CONNECTION_CHANGE message. This will add new devices and remove old ones.
2. We look at the device ID from the PUSH_DOPPLER_CONNECTION_CHANGE message and if the device does not exist in the devices list, then we trigger update_devices() (we'd still need to trigger update_devices() on any OFFLINE PUSH_DOPPLER_CONNECTION_CHANGE message.

**Comment by alandtse** [comment]
I think 2 is better. I did notice a lot of Doppler messages in my setup with 5 devices so I think it'd be a lot of unnecessary calls. I don't think we need to remove entities though since we already ignore unavailable players. I guess the question is whether Bluetooth changes can also be addressed by websocket. That is the last thing we're polling for I think.

**Comment by keatontaylor** [comment]
Do you use the Alexa app often? I know it does a Doppler state change for itself, which happens each time you open and close the app. Echo devices shouldn't be sending online/offline messages often unless there is a spotty connection from the echo's to amazon's servers.

**Comment by keatontaylor** [comment]
Also, I created a new issue and am working on adding bluetooth state change notifications. Also if we are not planing on removing entities until a reboot of HA, it is absolutely fine by me to implement adding players on a Doppler online event.

**Comment by alandtse** [comment]
Oh it might have been during testing. I left the web app on so that may have done it.

**Comment by alandtse** [comment]
FYI, further updates to my dev branch. 

As a note, I have a third case we didn't consider. Since we're monitoring multiple websocket messages that can identify a new `deviceSerialNumber`, I think it makes sense to always check it against the serialNumbers we already know. If it's not known, it will set a flag that will let `update_devices` do the rest API call despite the websocket.

I left the update_devices on a scanning_interval timer. It just will return immediately if websockets is connected unless `(hass.data[DATA_ALEXAMEDIA]['accounts'][email]['new_devices']) is True`
